### DISCUSSION TITLE
Types of functionals J_T

### DISCUSSION THREAD

**Discussion by kayelderson** [qucontrol/krotov#112] [Q&A] [answered]
I'm wondering why fidelity for density matrices is not included in the objective functional listed in the package. It seems to me at the moment the reason is because $-\partial J_T/ \partial \langle \langle \rho \rvert$ for fidelity $J_T = \text{tr}(\sqrt{\sqrt{\rho} \sigma \sqrt{\rho}})^2$ hard to compute. If so, I'd like to customize it, anyone has idea how to obtain $-\partial J_T/ \partial \langle \langle \rho \rvert$ in this case?

**Comment by goerz** [comment, marked as answer]
This package is definitely "bring your own functional".

You can use matrix calculus to obtain an analytic derivative and implement that; or you can try to use automatic differentiation (AD) to have the computer calculate it for you, see [Quantum 6, 871 (2022)](https://quantum-journal.org/papers/q-2022-12-07-871/). In Python, Jax might be appropriate for AD. I haven't tried Jax myself, but I hear good things.

The [Julia version of this package](https://github.com/JuliaQuantumControl/Krotov.jl) has support for automatic differentiation built in, so there, it should be sufficient to implement just `J_T`, and the optimization will calculate the gradient automatically. The Zygote library that does the AD for Julia isn't always completely reliable though, so I strongly recommend testing the AD-gradient for correctness by comparing it with finite differences.

**Reply by kayelderson** [reply]
Hi Goerz, thank you so much for your instantaneous, kind response, I will try the AD suggestion. Speaking of the analytic derivative, is it true that $\partial J_T/\partial \langle \langle \rho \lvert = \partial J_T/\partial \rho$? (If so, it seems like https://math.stackexchange.com/questions/2800228/derivative-of-mboxtr-left-leftaxa-right-frac12-right-with-respe answers the question).

**Reply by kayelderson** [reply]
The fidelity works properly using the analytic form. Thank you :)

**Reply by goerz** [reply]
It's a little tricky because the elements of ρ are complex-valued. So that notation combines a matrix derivative and a Wirtinger derivative. See Section 4.1 of the [Quantum article](https://quantum-journal.org/papers/q-2022-12-07-871/) I linked for how exactly this is written out. The Stackexchange post might be relevant here, but it's probably for real matrices, so be careful with that.

I think "traditionally" the Bures fidelity that you want to use has been considered as "not analytically differentiable" and instead, people have used a Hilbert-Schmidt distance. See [Adv. Quantum Tech. 2, 1800110 (2019)](https://onlinelibrary.wiley.com/doi/10.1002/qute.201800110) ([arXiv](https://arxiv.org/abs/1807.04718)). @Basilewitsch Can you elaborate?

**Reply by goerz** [reply]
Just to be fully explicit:

```math
-\frac{\partial J_T}{\partial \langle \langle \rho \vert\vert}
\equiv -\frac{1}{2} \sum_{nm} \left(
  \frac{\partial J_T}{\partial \text{Re}[\rho_{nm}]}
  + i \frac{\partial J_T}{\partial \text{Im}[\rho_{nm}]}
\right) \vert n \rangle \langle m \vert
```

for an arbitrary basis $n, m$. That's very tedious, of course, but the "matrix calculus" ultimately has to be compatible with that expression.

**Reply by kayelderson** [reply]
I believe $\partial J_T/\partial \langle \langle \rho \lvert = \partial J_T/\partial \rho$ is true in general, so that your derivative expression of using double-ket notation will eventually be equivalent to the derivative taken wrt $\rho$. But it's a good point that the Stackexchange post is only for real matrices, I will double check it.

I've been reading [Adv. Quantum Tech. 2, 1800110 (2019)](https://onlinelibrary.wiley.com/doi/10.1002/qute.201800110), and I think what you meant is described in a paragraph below Eq.(15) in that reference. I am very confused by the paragraph saying _"no closed analytical expression for their derivative"_ because [this post](https://math.stackexchange.com/questions/4616514/gradient-of-trace-distance-schatten-1-norm) gives a closed analytic derivative of the trace distance for Hermitian matrices.

**Reply by goerz** [reply]
I think this might be a case of "to the best of our knowledge", or maybe it should say "a closed analytical expression for their derivative is not straightforward to obtain". "No analytic derivative exists" may turn out to not strictly be true. But I don't think that matrix calculus, especially for complex matrices, is widely known. The only book I'm aware of that develops this systematically is ["Complex-Valued Matrix Derivatives" by Are Hjørungnes](https://www.cambridge.org/core/books/complexvalued-matrix-derivatives/12792AB48C0CA693F8591AAC95451A7F). I wouldn't have had any clue on how to formulate derivatives of the measures 11-14 in the [AQT paper](https://onlinelibrary.wiley.com/doi/10.1002/qute.201800110) until fairly recently (when writing the [Semi-AD paper](https://quantum-journal.org/papers/q-2022-12-07-871/)), and I'd still consider it highly non-trivial. I'd very much use automatic differentiation in this situation. 

@Basilewitsch Do you want to chime in here?

**Reply by Basilewitsch** [reply]
I agree with @kayelderson that [this post](https://math.stackexchange.com/questions/4616514/gradient-of-trace-distance-schatten-1-norm) seems to provide the correct analytical derivative of the trace distance. So maybe that would have been another good measure to benchmark in the [AQT paper](https://doi.org/10.1002/qute.201800110).

At the time of writing the paper, we were at least not aware of analytical forms for most of the distance measures listed in Eqs. (11)-(15) and in general I agree with @goerz that if analytical derivatives exist they are most likely non-trivial to compute. However, if you want to maximize the fidelity, respectively minimize the infidelity, I would say it is completely fine to minimize one of the other distance measures, e.g. the trace or Hilbert-Schmidt distance, as minimizing it will minimize the infidelity as well.

**Reply by kayelderson** [reply]
Great, thank you all for your kind responses. Indeed, I implemented trace distance by constructing the chi function following the post and it worked properly (for both closed and open quantum systems). So the conclusion is the same: trace distance can be properly implemented in Krotov's method, and furthermore I'm pretty optimistic that Uhlmann fidelity and other metrics listed in the AQT paper would also have analytic derivatives. I'll update here in case if I figure them out later.

**Reply by goerz** [reply]
(via @danielreich). Just to highlight a relevant section of the paper: "When resorting to numerical evaluation of the derivatives, the required spectral decomposition results in accuracy problems due to the presence of square roots, respectively logarithms, of the state. This problem becomes particularly severe in the common case that several eigenvalues of the density matrix are close to zero, rendering most of the aforementioned functionals impractical for application in gradient-based optimization."

So there is the presumption that you might run into numerical instabilities with the derivative of the Bures fidelity (or the other measures 11-15). It would be interesting to see what your experience is with this. Keep us up to date!

In general, though: is there a reason the Hilbert-Schmidt fidelity wouldn't work for you? It should definitely be much more well-behaved for optimization than any of the other functionals, and if you're trying to do open system state-to-state transitions, it should capture your intent.

**Reply by kayelderson** [reply]
@goerz I've found that the paper was totally correct in the sense that there is a huge numerical instability when using trace distance. For example of 2-qubit case, the below figure is an example of the chi function, which is extremely large. This is because $(X^2)^{-1/2}$ part in the gradient ([this post](https://math.stackexchange.com/questions/4616514/gradient-of-the-trace-distance-schatten-1-norm)) is extremely large due to very small eigenvalues of $X$. I believe the same numerical instability would appear when using other metrics since their analytic closed forms of their gradients (if exist) would also involve the inverse of a matrix that could have small eigenvalues. 
<img width="1000" alt="Screen Shot 2023-01-12 at 4 08 33 PM" src="https://user-images.githubusercontent.com/122472163/212200554-1910b241-9049-4b73-b015-b1616224378d.png">

Regarding your question: Hilbert-Schmidt distance would work for me! I was just wondering on why people wouldn't use trace distance/Uhlmann fidelity as they are the most popular metrics.
